### ISSUE TITLE
REPL crashes on Emacs 27

### ISSUE THREAD

**Issue by arthurcgusmao** [nnicandro/emacs-jupyter#219]
After switching to Emacs 27.0.60, I started to experience crashes/super slowness when interacting with a Jupyter REPL buffer. After pressing `C-g` several times to quit the crashing state, the following was output in the messages buffer:
```
Error during redisplay: (jit-lock-function 180) signaled (quit)
Error during redisplay: (jit-lock-function 181) signaled (quit)
Error during redisplay: (jit-lock-function 182) signaled (quit)
Error during redisplay: (jit-lock-function 186) signaled (quit)
Error during redisplay: (jit-lock-function 187) signaled (quit)
Error during redisplay: (jit-lock-function 193) signaled (quit)
Error during redisplay: (jit-lock-function 194) signaled (quit)
Error during redisplay: (jit-lock-function 180) signaled (beginning-of-buffer)
```

So it must have something to do with displaying the buffer, and, more precisely, with `font-lock`. Indeed, when I disable `font-lock` (such as when using the hotfix snippet below), the slowness/crashing stops. It doesn't seem to be anything particular to my Emacs configurations since it was working normally before. How should we fix it?

```lisp
;; Hotfix
(add-hook 'jupyter-repl-mode-hook
          (lambda () (font-lock-mode 0)))
```

EDIT
Maybe something to do with [this new feature](https://github.com/emacs-mirror/emacs/blob/master/etc/NEWS.27#L718-L726)?
I tried setting the variable `jit-lock-antiblink-grace` to `nil` but the behavior remained the same, with the same errors.

**Comment by alendit** [comment]
@dzop The override helped in my case (Emacs 27.0.90).

**Comment by iskur** [comment]
I cannot speek for @arthurcgusmao, but adding @dzop's override to my jupyter configuration solved a different problem for me (sorry, I did not open an issue for that earlier).
When entering an undefined variable and pressing enter in the jupyter REPL, emacs would freeze and become unresponsive even to repeated ctrl-g keypresses.
(Emacs 28.0.50)

**Comment by arthurcgusmao** [comment]
Thanks @dzop, the override helped in my case also. The issue in `jupyter-repl-font-lock-fontify-region` is on the line containing
```lisp
(funcall fontify-fun (point-min) (point-max) verbose)
```
, when I comment it out no freezing happens, otherwise, it does.

It tried to investigate it but it seems that the function is only called inside `jupyter-repl-initialize-fontification` and the argument `fontify-fun` is `font-lock-fontify-region-function`, a built-in function in `font-lock.el`. No idea how to further debug it, what would you suggest?

**Comment by dzop** [comment]
I suspect this has something to do with the narrowing that is implied by the call to `jupyter-repl-cell-cond`. The expansion of `jupyter-repl-cell-cond` in `jupyter-repl-font-lock-fontify-region` will reveal the narrowing:

```elisp
(defun jupyter-repl-font-lock-fontify-region (fontify-fun beg end &optional verbose)
  (save-excursion
    (save-restriction
      (let ((start beg)
            (-end end)
            next)
        (while (/= start -end)
          (widen)
          (cond ((eq (get-text-property start 'field) 'cell-code)
                 (setq next (min -end (field-end start t)))
                 (narrow-to-region start next)
                 (funcall fontify-fun (point-min) (point-max) verbose))
                (t (setq next (or (text-property-any start -end 'field 'cell-code)
                                  -end))
                   (narrow-to-region start next)
                   (font-lock-unfontify-region (point-min) (point-max))))
          (setq start next))))))
```

Maybe something has changed in `font-lock-default-fontify-region`? We narrow to the REPL input cell when fontifying so that none of the output is fontified by the `major-mode` of the language.

I'm still on Emacs 26.3 at the moment, but will upgrade soon. If someone can write a test case which causes the error, it would be a lot easier for me to see if I can think of anything else. For example, the kernel being used, contents of a REPL cell, position of point within the contents, and the key sequence that causes the error.

**Comment by arthurcgusmao** [comment]
Here is the diff for `font-lock.el` (Emacs 26.3 of the left and 27.0.90 on the right-hand side)
![image](https://user-images.githubusercontent.com/12454729/76671839-d0c18800-6598-11ea-8118-e1bbe7fa820f.png)

![image](https://user-images.githubusercontent.com/12454729/76672047-6dd0f080-659a-11ea-8e16-3cd1fa58fd04.png)


As we can see,
```lisp
(save-restriction
    (unless font-lock-dont-widen (widen))
    ...)
```
was moved from `font-lock-default-fontify-region` to `font-lock-fontify-region`, so it seems that calling the latter function instead of the former might solve it. I'm going to try and, if it works, send a pull request.



---
You can verify the whole diff in `font-lock.el` by running:
```console
git clone https://github.com/emacs-mirror/emacs.git && cd emacs
git diff emacs-26.3 emacs-27.0.90 -- **/font-lock.el
```
(Actually, I had to indent the file in Emacs 26.3 to make the diff more meaningful)

**Comment by arthurcgusmao** [comment]
@dzop, please notice that my fix introduces a call to `font-lock-set-defaults` (present in `font-lock-fontify-region`, the new function being used) that wasn't there before.

Any idea what this could interfere with? The function seems to use the value of `font-lock-defaults`, modified in `jupyter-repl-initialize-fontification`. Previously, where would `font-lock-defaults` be utilized?

**Comment by dzop** [comment]
It looks like `font-lock-defaults` gets used when evaluating the `font-lock-set-defaults` call in `font-lock-fontify-buffer`. `font-lock-fontify-buffer` is eventually called after the mode hooks are run in the call to `font-lock-mode` in `jupyter-repl-initialize-fontification`.

The `:after-hook` entry in the minor mode def. of `font-lock-mode` shows that, after the mode hooks are run, a call to `font-lock-initial-fontify` is made which in turn makes the call to `font-lock-fontify-buffer`.

Note,  I'm looking at the functions as they are in Emacs 26.3 so maybe somethings changed Emacs 27 in one of those functions too.

Instead of using `font-lock-fontify-region`, maybe make a call to `font-lock-set-defaults` after the call to `font-lock-mode` in `jupyter-repl-initialize-fontification`?

**Comment by arthurcgusmao** [comment]
Why making a call to `font-lock-set-defaults` instead of using `font-lock-fontify-region`? This by itself wouldn't solve the issue, I guess.

The issue started happening after, from Emacs-26 to Emacs-27, the code block
```lisp
(save-restriction
  (unless font-lock-dont-widen (widen))
  ...)
```
was refactored from `font-lock-default-fontify-region` to `font-lock-fontify-region`, hence the solution proposed of substituting the former by the latter (which I now realize is **not** the right thing to do, indeed). Anyways, to explain myself, my concern in the previous message was that the latter calls `font-lock-set-defaults`, something the former didn't, so we'd now be calling this function while in the past we didn't (and now I realize the function shouldn't be called there, even though it seems that no side-effects happen).

I was able to replicate the freezing in Emacs-26.3 by commenting out the line that performs widening of the region right below `save-restriction`, as shown below.
```lisp
(defun font-lock-default-fontify-region (beg end loudly)
  "Fontify the text between BEG and END.
If LOUDLY is non-nil, print status messages while fontifying.
This function is the default `font-lock-fontify-region-function'."
  (save-buffer-state
    ;; Use the fontification syntax table, if any.
    (with-syntax-table (or font-lock-syntax-table (syntax-table))
      (save-restriction
        ;; (unless font-lock-dont-widen (widen))  ; <-- Comment out this line and freeze
        ;; Extend the region to fontify so that it starts and ends at
        ;; safe places.
        (let ((funs font-lock-extend-region-functions)
              (font-lock-beg beg)
              (font-lock-end end))
          (while funs
            (setq funs (if (or (not (funcall (car funs)))
                               (eq funs font-lock-extend-region-functions))
                           (cdr funs)
                         ;; If there's been a change, we should go through
                         ;; the list again since this new position may
                         ;; warrant a different answer from one of the fun
                         ;; we've already seen.
                         font-lock-extend-region-functions)))
          (setq beg font-lock-beg end font-lock-end))
        ;; Now do the fontification.
        (font-lock-unfontify-region beg end)
        (when (and font-lock-syntactic-keywords
                   (null syntax-propertize-function))
          ;; Ensure the beginning of the file is properly syntactic-fontified.
          (let ((start beg))
            (when (< font-lock-syntactically-fontified start)
              (setq start (max font-lock-syntactically-fontified (point-min)))
              (setq font-lock-syntactically-fontified end))
            (font-lock-fontify-syntactic-keywords-region start end)))
        (unless font-lock-keywords-only
          (font-lock-fontify-syntactically-region beg end loudly))
        (font-lock-fontify-keywords-region beg end loudly)
        `(jit-lock-bounds ,beg . ,end)))))
```
(if testing with this, notice you have to eval the whole `font-lock.el` buffer for the function to be re-assigned properly)

Thus, it is the lack of widening of the region in `font-lock-default-fontify-region` in Emacs-27 that is causing the issue. Interestingly, this goes against what is written in the docstring of `jupyter-repl-font-lock-fontify-region`:
```lisp
;; Ensure that the buffer is narrowed to the actual cell code before
;; calling the REPL language's `major-mode' specific fontification
;; functions since those functions don't know anything about input cells
;; or output cells and may traverse cell boundaries.
```
It doesn't make a difference to narrow to the cell region if it is widened later; the only things that change are the region-delimiting arguments passed to the function, since they are evaluated using `(point-min)` and `(point-max)` after the region is narrowed. But, we could pass the same values without narrowing. Could you elaborate a bit on this @dzop?

I'm still trying to understand exactly what these font-lock functions do; sorry for possible misunderstandings.

PS1: please ignore my commit above, it is a fix that worked for me but since I'm widening the region before calling `(point-min)` and `(point-max)` I now see those values are not the correct ones that should be passed, so it's not a proper fix.

---
Regarding the test you mentioned previously, I suggest the following:
1. Open a new Emacs buffer on Python mode
2. Insert any Python code (e.g., `l = [1,2,3]`)
    - Literally, any code I tried sending to the REPL was resulting in this bug
3. Create and associate a REPL to the buffer (`jupyter-run-repl`)
4. Go back to the Python buffer and Eval the buffer content (`jupyter-eval-buffer`); position of point doesn't seem to matter
5. Emacs will freeze if the issue is present

Software versions:
```text
Emacs 27.0.90
Python 3.7.3 (default, Mar 27 2019, 22:11:17) 
IPython 7.8.0 -- An enhanced Interactive Python.
```

PS2: I tried understanding how the tests in `/tests` you created work but my understanding is a bit limited yet; it would be nice to have it as a coded test. Conversely, possibly the existing tests will already catch this bug if run with Emacs-27.

**Comment by dzop** [comment]
> Could you elaborate a bit on this @dzop?

You're right that the narrowing has no effect, my fault for not paying attention more.

I did think that something was fishy when I saw `font-lock-dont-widen` since I had never seen that variable before (even though I have looked at `font-lock-default-fontify-region`) and it defaults to `nil`  which means the narrowing we do before calling `font-lock-default-fontify-region` is undone within the call, as you have already explained.

We would like the narrowing to happen, so given that the documentation of `font-lock-dont-widen` says:

```
If non-nil, font-lock will work on the non-widened buffer.
Useful for things like RMAIL and Info where the whole buffer is not
a very meaningful entity to highlight.
```

We can just `let` bind `font-lock-dont-widen` to `t`  around the call below my comment that you referenced.

**Comment by xzz53** [comment]
My two cents: I bisected it and it it looks like the culprit is https://github.com/emacs-mirror/emacs/commit/a642b4d35c86b871a97a4d5f150ed17303603413. Factoring out whitespace changes, the diff boils down to just
```diff
diff --git a/lisp/font-lock.el b/lisp/font-lock.el
index c20b410a68..c44041da07 100644
--- a/lisp/font-lock.el
+++ b/lisp/font-lock.el
@@ -1071,7 +1071,9 @@ font-lock-fontify-region
 If LOUDLY is non-nil, print status messages while fontifying.
 This works by calling `font-lock-fontify-region-function'."
   (font-lock-set-defaults)
-  (funcall font-lock-fontify-region-function beg end loudly))
+  (save-restriction
+    (unless font-lock-dont-widen (widen))
+    (funcall font-lock-fontify-region-function beg end loudly)))
 
 (defun font-lock-unfontify-region (beg end)
   "Unfontify the text between BEG and END.
@@ -1221,8 +1223,6 @@ font-lock-default-fontify-region
   (save-buffer-state
    ;; Use the fontification syntax table, if any.
    (with-syntax-table (or font-lock-syntax-table (syntax-table))
-      (save-restriction
-        (unless font-lock-dont-widen (widen))
      ;; Extend the region to fontify so that it starts and ends at
      ;; safe places.
      (let ((funs font-lock-extend-region-functions)
@@ -1251,7 +1251,7 @@ font-lock-default-fontify-region
      (unless font-lock-keywords-only
        (font-lock-fontify-syntactically-region beg end loudly))
      (font-lock-fontify-keywords-region beg end loudly)
-        `(jit-lock-bounds ,beg . ,end)))))
+     `(jit-lock-bounds ,beg . ,end))))
 
 ;; The following must be rethought, since keywords can override fontification.
 ;;    ;; Now scan for keywords, but not if we are inside a comment now
```

**Comment by arthurcgusmao** [comment]
@xzz53 yes, exactly. The problem now is how to adapt `emacs-jupyter` to that refactoring. Before the refactoring (Emacs-26), the buffer was being widened before executing the rest of the body of `font-lock-unfontify-region`, each time this function was called by `jupyter-repl-font-lock-fontify-region` (more specifically, by the macro `jupyter-repl-cell-cond` that `jupyter-repl-font-lock-fontify-region` calls). In Emacs-27, the buffer is not widened anymore in each of these calls and this is specifically why the bug is happening.

@dzop the problem is not making the narrowing take place (the narrowing is currently happening in `emacs-jupyter` in Emacs-27, whereas it wasn't in Emacs-26), but the narrowing itself is what is causes the bug to happen. If you enable narrowing in Emacs-26 as you suggested, the bug will appear.

So what is left for investigation is:

- Why narrowing causes the bug
- How to solve the issue in the best possible manner

**Comment by dzop** [comment]
After looking into it further, I found an un-terminating loop related to the
handling of `font-lock-extend-region-functions`
in `font-lock-default-fontify-function`.  Maybe this is what is making Emacs
freeze in your case as well.  Here is a test that shows what I am talking about

```elisp
    (ert-deftest jupyter-repl-issue-219 ()
      :tags '(repl i219)
      (jupyter-test-with-python-repl client
    	(jupyter-with-repl-buffer client
    	  (jupyter-repl-replace-cell-code "l = [1,2,3]")
    	  (jupyter-test-repl-ret-sync)
    	  (let* ((count 0)
    			 (font-lock-extend-region-functions 
    			  (cons (lambda ()
    					  (when (> (setq count (1+ count)) 100)
    						(error "Un-terminating `font-lock-extend-region-functions' handling"))
    					  nil)
    					font-lock-extend-region-functions)))
    		(font-lock-ensure)))))
```

You can add the test somewhere in `test/jupyter-test.el` and in the top level
directory of this project

1.  `M-x compile RET make dev RET` (only once)
2.  `M-x compile RET make test TAGS=i219 RET`

**NOTE:** You will need to have [Cask](https://github.com/cask/cask) installed.
  Also, you will need a copy of this project that includes the `test/`
  directory.  That directory isn't included in the MELPA package.

Finding that un-terminating loop led me to a bug in the
way `jupyter-repl-without-continuation-prompts` was implemented.  I've gone ahead
and fixed it in b691d38483b6540d42d482a32d35eb54178e5658.  Maybe this issue is
also fixed?  After fixing the underlying problem
with `jupyter-repl-without-continuation-prompts` I was able to get the above test
to pass.

**Comment by xzz53** [comment]
I confirm the issue is fixed for me in b691d38483b6540d42d482a32d35eb54178e5658, on Emacs 27.0.90. Thanks!

**Comment by arthurcgusmao** [comment]
Awesome! [b691d38](https://github.com/dzop/emacs-jupyter/commit/b691d38483b6540d42d482a32d35eb54178e5658) solved it for me as well, on Emacs 27.0.60. Thank you for the great work @dzop.

**Comment by arthurcgusmao** [comment]
Unfortunately, it seems that this bug persists. I started to experience it again when evaluating larger blocks of code, such as when using `jupyter-eval-buffer`. The crash stops happening when `font-lock-mode` is disabled at the jupyter-repl buffer, thus it points to this issue.

In trying to investigate it, I found that there is (on my machine at least) a limit of 497 characters in the Python code that we can evaluate. If the number of characters is greater than that, then the issue starts happening.

The *content* of the code doesn't seem to matter. For example, a code block like
```python
print("This code block has 498 chars and will make REPL crash")
print("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
print("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
print("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
print("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
print("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
print("aaaaa")
```
causes a crash just as one like
```python
# this line has 498 chars and will make the REPL crash aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
So makes it crash is the Python code that is trying to fontified in the REPL buffer when that code is sent to it, and not the output or computation of such code.

The more characters after the 497 threshold, the longer it takes to quit from frozen behavior of Emacs by pressing <key>C-g</key>, *if* you manage to restore it at all.

**Comment by Xparx** [comment]
emacs: `GNU Emacs 28.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.22.30, cairo version 1.15.10) of 2020-06-28`

I have this issue as well. I'm working with a pandas DataFrame (`291214 rows × 6 columns`).

```
target | regulator | combined_confidences | gold_standard | precision | recall
-- | -- | -- | -- | -- | --
BSU37040 | BSU31530 | 1.000e+00 | 1 | 1.000 | 3.181e-04
BSU34240 | BSU24610 | 1.000e+00 | 1 | 1.000 | 6.361e-04
BSU03340 | BSU25100 | 1.000e+00 | 1 | 1.000 | 9.542e-04
BSU34300 | BSU24610 | 1.000e+00 | 1 | 1.000 | 1.272e-03
```
printing the data works fine but (trying) manipulating the dataframe in any way freezes emacs.

I get the following in my Message buffer after pressing `C-g` repeatadly. 
```
Error during redisplay: (jit-lock-function 902) signaled (quit)
Error during redisplay: (jit-lock-function 903) signaled (quit)
Error during redisplay: (jit-lock-function 904) signaled (quit)
```

If I manage to get to `*Message*` emacs seems to be ok. If I go back to my repl. It freezes again.

Is there a way I can mitigate (removing functionality) this until the bug is fixed?

**Comment by Xparx** [comment]
I just saw the "Hot fix" in the first comment. It seems to be working.

**Comment by matteha** [comment]
Facing the same problem with freezing/crashing when large blocks of code are sent to the REPL (using jupyter-eval-region). The hot fix 'works' but also gets rid of all the fonts so the REPL buffer is just all black and white.

**Comment by astoff** [comment]
It looks like nobody is available at the moment to address this rather serious issue. Should the hot fix of the first comment be temporarily included in the package?

**Comment by nnicandro** [comment]
I've made an attempt at fixing this issue, see https://github.com/nnicandro/emacs-jupyter/pull/308. Can someone try it out? I used the code in @arthurcgusmao's comment https://github.com/nnicandro/emacs-jupyter/issues/219#issuecomment-641487092.

**Comment by astoff** [comment]
Works for me!

**Comment by timlod** [comment]
Works here as well, 27.1!

**Comment by nnicandro** [comment]
Fixed by #308.
